### DISCUSSION TITLE
Is it possible for Leptos CSR to generate static HTML for export?

### DISCUSSION THREAD

**Discussion by englut** [leptos-rs/leptos#3039] [Q&A] [answered]
I'm currently using Leptos in CSR mode and I'm wondering if it's possible to export my pages to HTML (static site)? 

I really like the react-like syntax that leptos provides. And the build tools are really great. But with what I'm currently building, WASM may be overkill. Was wondering if there's a way to have leptos export to HTML. 

I see that there's SSG on the SSR side of Leptos. I'm not sure if this just generates static HTML and serves it from the server-side binary? Or if leptos SSG actually generates static HTML files that can be served directly via CDNs and other static site hosts.

**Comment by gbj** [comment, marked as answer]
I think there's maybe a conceptual mismatch here.

For the sake of clarifying terms:
- CSR = client-side rendering. Runs code in the browser to rendering your application to DOM nodes, by compiling Rust code to WebAssembly (WASM)
- SSR = server-side rendering. Runs code anywhere for rendering your application to HTML, by compiling Rust code to a whatever native binary format you'd like. "Server side" is something of a misnomer, as it does not require a "server" per se. Rather, this is "rendering HTML in a binary."

So "I'd like to use CSR without WASM" is just not possible.

However, I think what you're asking is whether you can use static site generation without actually running a server. The answer to this is "yes, absolutely."

If you:
- start with the static routing example in the current git main (this is 0.7.0-beta)
- remove the `HydrationScripts` component
- remove the parts where you're actually starting up an Axum server, but
- keep the parts where it's generating static files

You should effectively be able to run this with `cargo run --features=ssr` and it will just build your site to static HTML files in `/target/site`.

**Reply by englut** [reply]
@gbj Thanks for this. I appreciate a clear answer on "CSR without WASM." That's fascinating that we can do "SSR" in the client! 

Having separate HTML+WASM files is close to what I want, and is good enough for what I am trying to do 🙏

Ah, it's a 0.7 feature. I was wondering why something like this wasn't in the leptos book. 

I'm trying to make it work, but I think I have to read more about Axum and Leptos hydration feature, before I know what to keep and what to remove. Maybe we can get a working example added?

**Reply by gbj** [reply]
> Having separate HTML+WASM files is close to what I want, and is good enough for what I am trying to do 🙏

I'm not sure I understand -- I thought you said you wanted not to use WASM?

Or are you just looking for static site generation in general? In that case, it's even easier. Just remove the part of the static routing example where it actually starts a server, and it will generate all the pages as HTML and you can serve the `target/site` directory. I'm saying just removing [these lines](https://github.com/leptos-rs/leptos/blob/1b48a2f8d550f52c47ba2372c8608f0b3a99dad8/examples/static_routing/src/main.rs#L20-L34) should be what you're looking for. I don't think this needs a separate example.

**Reply by englut** [reply]
My bad. I had worded my ask as if it was one thing. But I basically had two questions. First was whether or not it's possible to export to separate HTML files. I was having issues with Sourcehut pages needing a separate html file for the 404 page.

I just tested your suggestion, and it works how I need it to work. Thank you so much! I wasn't sure if I needed to remove more or not. But good to know that it's just literally removing the server running aspect. You're right, definitely don't need a separate example for this. I haven't seen any documentation for the static routes yet (I'm sure it will show up when 0.7 is released!), but will be good to have in there a little blurb about removing the server and running `cargo run --features=ssr` can export static routes to separate HTML files 🙏

The second question was me thinking theoretically. If static site generation can be exported, it'd _only be_ (famous last words) another step to export the HTML that was generated. But I understand that it's a lot more easier (and probably out of scope) to theorize this than to do it. So I am glad to get the clear answer you gave me that it's not possible with leptos.

That's all to say... we're aligned! Thank you for helping me out. I really appreciate it!

**Reply by gbj** [reply]
> If static site generation can be exported, it'd only be (famous last words) another step to export the HTML that was generated.

I don't think I understand this sentence. What does it mean to export the HTML that was generated? This should all be saved as `.html` files in `target/site`: Do you mean something else by "exporting"?

**Reply by englut** [reply]
Ah yes! I'm using "export" the way you're using "saved as .html." 

I used the term "export", because a feature I've used in the past from Next.js is called [static exports](https://nextjs.org/docs/pages/building-your-application/deploying/static-exports), which is similar to what I'm asking for here.

**Reply by englut** [reply]
Oh dang. I had thought the `target/sites` HTML files were still including WASM, but they're exactly what I was looking for. My bad for the confusion @gbj.
